### PR TITLE
Fail Build Team Deprecated PowerShell Tasks

### PR DESCRIPTION
**Pull request by Pranshu-Negi**
**Task name**: ChefKnifeV1, ChefV1, IISWebAppDeployment, QuickPerfTestV1, RunJMeterLoadTestV1, RunLoadTestV1, SqlServerDacpacDeployment

**Description**: Fail Deprecated Tasks with removal date 31-01-2024 owned by Agent and Tasks team 

**Documentation changes required:** N

**Added unit tests:** (Y/N)

**Attached related issue:** https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/2125781

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected

### CODE REVIEW

**Review by DmitriiBobreshev — APPROVED** [review]

File: Tasks/ChefKnifeV1/ChefKnife.ps1 [hunk 1]
```diff
@@ -31,6 +31,13 @@ try
     #setting up chef repo with the chef subscription details fetched before 
     Initialize-ChefRepo $connectedServiceDetails 
     Invoke-Expression -Command $scriptCommand
+    $featureFlags = @{
+        failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+    }
+	if ($featureFlags.failDeprecatedBuildTask)
+	{
+		throw "The ChefKnife@1 (Run scripts with Knife commands on your Chef workstation) task has been deprecated since March 5, 2018 and will soon be retired. To continue to use Chef Knife, use the Knife client directly from a bash/pwsh/script task. See https://docs.chef.io/workstation/knife_client/."
```
Comment (geekzter) [inline Tasks/ChefKnifeV1/ChefKnife.ps1:39]: Shouldn't this string be sourced from task.json?

File: Tasks/ChefV1/Chef.ps1 [hunk 2]
```diff
@@ -186,6 +186,14 @@ try
 	Invoke-Knife @("upload environments/$environment.json")
 
 	Wait-ForChefNodeRunsToComplete $environment $totalWaitTimeForRunsInMinutes $pollIntervalForRunsInSeconds
+
+	$featureFlags = @{
+        failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+    }
+	if ($featureFlags.failDeprecatedBuildTask)
+	{
+		throw "The Chef@1 (Deploy to Chef environments by editing environment attributes) task has been deprecated since March 5, 2018 and will soon be retired. To continue to use Chef, use the Chef CLI directly from a bash/pwsh/script task. See https://github.com/chef/chef-cli."
```
Comment (geekzter) [inline Tasks/ChefV1/Chef.ps1:195]: Shouldn't this string be sourced from task.json?

File: Tasks/IISWebAppDeployment/DeployIISWebApp.ps1 [hunk 3]
```diff
@@ -154,4 +154,12 @@ if(-not [string]::IsNullOrEmpty($errorMessage))
     throw "$errorMessage $helpMessage"
 }
 
+$featureFlags = @{
+    failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+}
+if ($featureFlags.failDeprecatedBuildTask)
+{
+	throw "The IISWebAppDeployment@1 (IIS Web App deployment) task has been deprecated since October 5, 2018 and will soon be retired. Use the 'IIS Web App Deployment Using WinRM' extension instead: https://marketplace.visualstudio.com/items?itemName=ms-vscs-rm.iiswebapp."
```
Comment (geekzter) [inline Tasks/IISWebAppDeployment/DeployIISWebApp.ps1:162]: Shouldn't this string be sourced from task.json?

File: Tasks/QuickPerfTestV1/Invoke-QuickPerfTest.ps1 [hunk 4]
```diff
@@ -187,6 +187,13 @@ else
 	Write-Error ("Failed to connect to the endpoint '{0}' for VSO account '{1}'" -f $EndpointName, $VSOAccountUrl)
 }
 
+$featureFlags = @{
+	failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+}
+if ($featureFlags.failDeprecatedBuildTask)
+{
+	throw "The QuickPerfTest@1 (Cloud-based web performance test) task has been deprecated since June 4, 2019 and will soon be retired. Use the AzureLoadTest@1 task instead."
```
Comment (geekzter) [inline Tasks/QuickPerfTestV1/Invoke-QuickPerfTest.ps1:195]: Shouldn't this string be sourced from task.json?

File: Tasks/RunJMeterLoadTestV1/Start-ApacheJMeterTest.ps1 [hunk 5]
```diff
@@ -184,5 +184,12 @@ else
 	("Connection '{0}' failed for service '{1}'" -f $connectedServiceName, $connectedServiceDetails.Url.AbsoluteUri) >> $summaryFile
 }
 
+$featureFlags = @{
+	failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+}
+if ($featureFlags.failDeprecatedBuildTask)
+{
+	throw "The ApacheJMeterLoadTest@1 (Cloud-based Apache JMeter load test) task haas been deprecated since June 4, 2019 and will soon be retired. For cloud-based load testing, use the AzureLoadTest@1 task."
```
Comment (geekzter) [inline Tasks/RunJMeterLoadTestV1/Start-ApacheJMeterTest.ps1:192]: Shouldn't this string be sourced from task.json?

File: Tasks/RunLoadTestV1/Start-CloudLoadTest.ps1 [hunk 6]
```diff
@@ -233,5 +233,13 @@ else
 	Write-Error ("Connection '{0}' failed for service '{1}'" -f $connectedServiceName, $connectedServiceDetails.Url.AbsoluteUri)
 }
 
+$featureFlags = @{
+	failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+}
+if ($featureFlags.failDeprecatedBuildTask)
+{
+	throw "The CloudLoadTest@1 (Cloud-based load test) has been deprecated since June 4, 2019 and will soon be retired. Use the AzureLoadTest@1 task instead."
```
Comment (geekzter) [inline Tasks/RunLoadTestV1/Start-CloudLoadTest.ps1:241]: Shouldn't this string be sourced from task.json?

File: Tasks/SqlServerDacpacDeployment/DeployToSqlServer.ps1 [hunk 7]
```diff
@@ -81,4 +81,12 @@ if(-not [string]::IsNullOrEmpty($errorMessage))
     throw "$errorMessage $helpMessage"
 }
 
+$featureFlags = @{
+    failDeprecatedBuildTask  = [System.Convert]::ToBoolean($env:FAIL_DEPRECATED_BUILD_TASK)
+}
+if ($featureFlags.failDeprecatedBuildTask)
+{
+	throw "The SqlServerDacpacDeployment@1 (SQL Server database deploy) task has been deprecated since October 5, 2018 and will soon be retired. Use the 'IIS Web App Deployment Using WinRM' extension instead: https://marketplace.visualstudio.com/items?itemName=ms-vscs-rm.iiswebapp. For Azure SQL Database use the SqlAzureDacpacDeployment@1 task."
```
Comment (geekzter) [inline Tasks/SqlServerDacpacDeployment/DeployToSqlServer.ps1:89]: Shouldn't this string be sourced from task.json?

**Review by geekzter — CHANGES_REQUESTED** [review]

File: Tasks/ChefKnifeV1/ChefKnife.ps1 (hunk 1 above)
Comment (Pranshu-Negi) [reply Tasks/ChefKnifeV1/ChefKnife.ps1:39]: For PowerShell based tasks, in order to read the strings from resource file generated from task.json, we need to add another module (VstsTaskSdk) in the dependencies of the task. For a deprecated Task, I believe we shouldn't add a new dependency.

File: Tasks/ChefV1/Chef.ps1 (hunk 2 above)
Comment (Pranshu-Negi) [reply Tasks/ChefV1/Chef.ps1:195]: Same as above

File: Tasks/IISWebAppDeployment/DeployIISWebApp.ps1 (hunk 3 above)
Comment (Pranshu-Negi) [reply Tasks/IISWebAppDeployment/DeployIISWebApp.ps1:162]: Same as above

File: Tasks/QuickPerfTestV1/Invoke-QuickPerfTest.ps1 (hunk 4 above)
Comment (Pranshu-Negi) [reply Tasks/QuickPerfTestV1/Invoke-QuickPerfTest.ps1:195]: Same as above

File: Tasks/RunJMeterLoadTestV1/Start-ApacheJMeterTest.ps1 (hunk 5 above)
Comment (Pranshu-Negi) [reply Tasks/RunJMeterLoadTestV1/Start-ApacheJMeterTest.ps1:192]: Same as above

File: Tasks/RunLoadTestV1/Start-CloudLoadTest.ps1 (hunk 6 above)
Comment (Pranshu-Negi) [reply Tasks/RunLoadTestV1/Start-CloudLoadTest.ps1:241]: Same as above

**Review by geekzter — APPROVED** [review]

**Review by DergachevE — APPROVED** [review]

**Pull request merged**